File: knapsack/LC6987.cpp
#include <bits/stdc++.h>
// https://leetcode.cn/problems/minimum-time-to-make-array-sum-at-most-x/
using namespace std;
#define i64 long long

class Solution {
public:
    int minimumTime(vector<int>& nums1, vector<int>& nums2, int x) {
        int n = nums1.size();
        vector<int> ids(n);
        iota(ids.begin(), ids.end(), 0);
        sort(ids.begin(), ids.end(), [&](int x, int y) {
            return nums2[x] < nums2[y];
        });
        vector<int> dp(n+1, 0);
        for(int i=0;i<n;++i){
            int num1 = nums1[ids[i]];
            int num2 = nums2[ids[i]];
            for (int j=n;j>0;--j){
                dp[j] = max(dp[j-1] + num1 + num2*j, dp[j]);
            }
        }
        int s1 = accumulate(nums1.begin(), nums1.end(), 0);
        int s2 = accumulate(nums2.begin(), nums2.end(), 0);
        for(int t=0;t<=n;++t){
            if (s1 + t*s2 - dp[t] <= x) return t;
        }
        return -1;
    }
};

int main(){
    return 0;
}

File: knapsack/LC6987.py
# https://leetcode.cn/problems/minimum-time-to-make-array-sum-at-most-x/
from typing import *


class Solution:
    # dp[][]
    def minimumTime(self, nums1: List[int], nums2: List[int], x: int) -> int:
        n = len(nums1)
        arr = sorted((nums2[i], nums1[i]) for i in range(n))
        # fix the elements to operate on, operate on index with larger nums2 last
        dp = [[0]*(n+1) for _ in range(n+1)]
        # dp[i][j]: deducted amount: considering the first i elements, after j operations (j <= n)
        for i in range(n):
            for j in range(1, n+1):
                num2, num1 = arr[i]
                # operating on nums1[i] VS not
                dp[i+1][j] = max(dp[i][j-1] + num1 + num2*j, dp[i][j])
        s1 = sum(nums1)
        s2 = sum(nums2)
        cur = s1
        for t in range(n+1):
            if cur - dp[n][t] <= x:
                return t
            cur += s2
        return -1


class Solution:
    # dp[]
    def minimumTime(self, nums1: List[int], nums2: List[int], x: int) -> int:
        n = len(nums1)
        arr = sorted((nums2[i], nums1[i]) for i in range(n))
        # fix the elements to operate on, operate on index with larger nums2 last
        dp = [0]*(n+1)
        # dp[i][j]: deducted amount: considering the first i elements, after j operations (j <= n)
        for num2, num1 in arr:
            for j in range(n, 0, -1):
                # operating on nums1[i] VS not
                dp[j] = max(dp[j-1] + num1 + num2*j, dp[j])
        s1 = sum(nums1)
        s2 = sum(nums2)
        cur = s1
        for t in range(n+1):
            if cur - dp[t] <= x:
                return t
            cur += s2
        return -1
